crates/oci-distribution/src/client.rs

```diff
@@ -411,7 +494,7 @@ mod test {
 
     #[tokio::test]
     async fn test_auth() {
-        let image = Reference::try_from(HELLO_IMAGE).expect("failed to parse reference");
+        let image = Reference::try_from(HELLO_IMAGE_TAGGED).expect("failed to parse reference");
```

Comment:
Can we add a separate test instead for HELLO_IMAGE_TAGGED? That way we can continue to test image references with no tag.

Reply:
In fact... It might be better to add all three into a matrix, such that each `test_*` function tests untagged, tagged, and digest references to make sure all three work in a consistent fashion.

Reply:
I took a crack at this in a8352eb. I'm not too familiar with table driven testing patterns in rust so my approach is probably a little naïve. Notably, when any of the test cases trigger a panic during a loop like the one below, later test cases are skipped.

https://github.com/deislabs/krustlet/blob/a8352ebe797eaa21d22778d1c0efb384f4151cd0/crates/oci-distribution/src/client.rs#L577-L587

It seems like we may need to use a macro to generate test functions in order to avoid this problem, but I'm not sure if this is desirable given the additional complexity that would incur. Are there any other approaches that might work? This [SO question](https://stackoverflow.com/questions/34662713/how-can-i-create-parameterized-tests-in-rust) has some interesting discussion but may be outdated.

Reply:
I just had some fun going down this rabbit hole. 😄 A refactor like this is probably best for a separate PR, but I came up with two macro based solutions, each resulting in identical cargo test output.

1. Define our own macro for each test type using [macro_rules](https://doc.rust-lang.org/rust-by-example/macros.html#macro_rules) (no dependencies)

```rust
macro_rules! test_to_v2_manifest {
    ($($name:ident: $value:expr,)*) => {
        mod to_v2_manifest {
            use super::*;
            $(
                #[test]
                fn $name() {
                    let (input, expected) = $value;
                    let reference = Reference::try_from(input).expect("failed to parse reference");
                    assert_eq!(Client::default().to_v2_manifest_url(&reference), expected);
                }
            )*
        }
    }
}

test_to_v2_manifest! {
    no_tag: (
        HELLO_IMAGE_NO_TAG,
        "https://webassembly.azurecr.io/v2/hello-wasm/manifests/latest",
    ),
    tag: (
        HELLO_IMAGE_TAG,
        "https://webassembly.azurecr.io/v2/hello-wasm/manifests/v1",
    ),
    digest: (
        HELLO_IMAGE_DIGEST,
        "https://webassembly.azurecr.io/v2/hello-wasm/manifests/sha256:51d9b231d5129e3ffc267c9d455c49d789bf3167b611a07ab6e4b3304c96b0e7",
    ),
    tag_and_digest: (
        HELLO_IMAGE_TAG_AND_DIGEST,
        "https://webassembly.azurecr.io/v2/hello-wasm/manifests/sha256:51d9b231d5129e3ffc267c9d455c49d789bf3167b611a07ab6e4b3304c96b0e7",
    ),
}
```

2. Use the [test_case](https://docs.rs/test-case/1.0.0/test_case/) crate (downside: it doesn't currently work with `#[tokio::test]`)

```rust
#[test_case(
    HELLO_IMAGE_NO_TAG,
    "https://webassembly.azurecr.io/v2/hello-wasm/manifests/latest";
    "no tag"
)]
#[test_case(
    HELLO_IMAGE_TAG,
    "https://webassembly.azurecr.io/v2/hello-wasm/manifests/v1";
    "tag"
)]
#[test_case(
    HELLO_IMAGE_DIGEST,
    "https://webassembly.azurecr.io/v2/hello-wasm/manifests/sha256:51d9b231d5129e3ffc267c9d455c49d789bf3167b611a07ab6e4b3304c96b0e7";
    "digest"
)]
#[test_case(
    HELLO_IMAGE_TAG_AND_DIGEST,
    "https://webassembly.azurecr.io/v2/hello-wasm/manifests/sha256:51d9b231d5129e3ffc267c9d455c49d789bf3167b611a07ab6e4b3304c96b0e7";
    "tag and digest"
)]
fn to_v2_manifest(image: &str, manifest_uri: &str) {
    let reference = Reference::try_from(image).expect("failed to parse reference");
    assert_eq!(
        Client::default().to_v2_manifest_url(&reference),
        manifest_uri
    );
}
```